File: ubf_app/lib/core/constants/admin_scopes.dart
import 'package:flutter/material.dart';
import 'package:mana/l10n/app_localizations.dart';

/// 공동 관리자가 맡는 분야 (059)
///
/// 서버의 SCOPES 와 **이름이 하나하나 같아야 한다**. 어긋나면 화면에서 고른
/// 분야가 서버에서 버려지고, 담당자는 골랐는데 안 된다고 여긴다.
/// 그것을 test/admin_scopes_test.dart 가 지킨다.
class AdminScope {
  final String key;
  final IconData icon;
  const AdminScope(this.key, this.icon);

  String label(AppLocalizations l) => switch (key) {
    'transport' => l.scopeTransport,
    'rooms' => l.scopeRooms,
    'groups' => l.scopeGroups,
    'ledger' => l.scopeLedger,
    'service' => l.scopeService,
    'registration' => l.scopeRegistration,
    'comms' => l.scopeComms,
    'schedule' => l.scopeSchedule,
    'medical' => l.scopeMedical,
    _ => key,
  };

  String hint(AppLocalizations l) => switch (key) {
    'transport' => l.scopeTransportHint,
    'rooms' => l.scopeRoomsHint,
    'groups' => l.scopeGroupsHint,
    'ledger' => l.scopeLedgerHint,
    'service' => l.scopeServiceHint,
    'registration' => l.scopeRegistrationHint,
    'comms' => l.scopeCommsHint,
    'schedule' => l.scopeScheduleHint,
    'medical' => l.scopeMedicalHint,
    _ => '',
  };
}

const adminScopes = <AdminScope>[
  AdminScope('transport', Icons.airport_shuttle_outlined),
  AdminScope('rooms', Icons.meeting_room_outlined),
  AdminScope('groups', Icons.menu_book_outlined),
  AdminScope('ledger', Icons.account_balance_wallet_outlined),
  AdminScope('service', Icons.volunteer_activism_outlined),
  AdminScope('registration', Icons.how_to_reg_outlined),
  AdminScope('comms', Icons.campaign_outlined),
  AdminScope('schedule', Icons.event_outlined),
  AdminScope('medical', Icons.medical_services_outlined),
];

/// 서버가 준 값 → 화면이 쓰는 집합.
///
/// **비어 있거나 없으면 전부다.** 059 이전에 세운 사람이 그렇고, 그 사람의
/// 권한을 화면이 마음대로 좁히면 안 된다.
Set<String>? scopesOf(Object? raw) {
  if (raw is! List || raw.isEmpty) return null;
  final s = raw.map((e) => '$e').toSet();
  return s.contains('all') ? null : s;
}

/// 이 분야를 볼 수 있는가. null 이면 전부라는 뜻이다.
bool canSee(Set<String>? mine, String scope) =>
    mine == null || mine.contains(scope);

File: server/src/services/admin_scopes.js
// 공동 관리자가 맡은 분야 — 순수 부분.
//
// **여기에는 DB 를 부르지 않는다.** auth.js 는 db.js 를 불러오고, db.js 는
// 불러오는 순간 DATABASE_URL 을 요구한다. 그래서 이 목록을 auth.js 에 두면
// 단위 테스트가 DB 없이는 뜨지 못하고, .env 가 있는 개발자 기계에서만
// 통과한다 — CI 에서는 처음부터 끝까지 빨간불이었다.

// 이름은 화면이 갈라지는 자리를 따라 지었다. 라우트마다 requireScope 로
// 어느 분야인지 적어야 하고, **안 적으면 검사가 막는다**(admin_scopes.test.js)
// — 안 적힌 라우트는 조용히 아무나 통과하기 때문이다.
export const SCOPES = [
  'transport',    // 픽업·교통·배차
  'rooms',        // 숙소
  'groups',       // 말씀공부 조
  'ledger',       // 장부·참가비
  'service',      // 봉사
  'registration', // 등록·명단
  'comms',        // 공지·자료실
  'schedule',     // 일정
  'medical',      // 의료·안전(SOS·질병 정보)
];

/// 이 사람이 이 수양회에서 맡은 분야.
/// null = 전부 · false = 권한 없음 · 배열 = 그 분야만.
export function requireScope(...needed) {
  return (req, res, next) => {
    const mine = req.adminScopes;
    if (mine === null || mine === undefined) return next(); // 전부
    if (needed.some((n) => mine.includes(n))) return next();
    return res.status(403).json({ error: '이 분야는 맡지 않으셨습니다' });
  };
}
